File: custom_scripts/custom_scripts/custom/js/purchase_order.js
frappe.ui.form.on('Purchase Order Item', {
	items_add: function(frm){
		frm.set_df_property('po_itemwise_rate_details', 'hidden', 1);
	},
  item_code: function (frm, cdt, cdn) {
    var d = locals[cdt][cdn]
    const set_fields = ['rate','date','supplier','purchase_order'];
    if(!d.item_code){
      frm.set_df_property('po_itemwise_rate_details', 'hidden', 1);
    }
    if(d.item_code){
      frm.set_df_property('po_itemwise_rate_details', 'hidden', 0);
      frappe.call({
        method: "custom_scripts.custom_scripts.custom.py.purchase_order.fetch_rate_details",
        args: {
          item_code: d.item_code
        },
        freeze: true,
        callback: function (r) {
            if(r.message) {
                frm.set_value('po_itemwise_rate_details', []);
                $.each(r.message, function(i, d) {
                    var row = frm.add_child('po_itemwise_rate_details');
                    for (let key in d) {
                        if (d[key] && in_list(set_fields, key)) {
                            row[key] = d[key];
                        }
                    }
                });
            }
            refresh_field('po_itemwise_rate_details');
        }
      })
    }
  }
});

File: custom_scripts/custom_scripts/custom/py/purchase_order.py
import frappe

@frappe.whitelist()
def fetch_rate_details(item_code):
    doc_count = 0
    rate_details = []
    po_details = frappe.get_all('Purchase Order Item',['rate','parent'],{'item_code':item_code,'parenttype':'Purchase Order'},order_by="modified")
    for row in po_details[::-1]:
        if frappe.db.get_value('Purchase Order', row.parent,'docstatus') == 1:
            po_doc = frappe.get_doc('Purchase Order', row.parent)
            rate_details.append(
                {
                'purchase_order': row.parent,
                'date': po_doc.transaction_date,
                'supplier': po_doc.supplier, 
                'rate': row.rate}
            )
            doc_count += 1
        if doc_count == 5:
            break
    return rate_details